File: lib/src/transformer.dart
/// Error → Problem transformer (C0 §2).
///
/// Folds an arbitrary thrown/returned value into a typed [Problem]. The HTTP →
/// Result/Problem bridge (`fromHttpError`) lives in `diene_api_engine`, not
/// here; this lib owns only the generic object → envelope fold used everywhere
/// a Dart value needs to become a Problem.
library;

import 'problem.dart';
import 'problem_type_uri.dart';
import 'registry.dart';

/// Options for folding an unknown value into a [Problem].
final class TransformOptions {
  /// Creates transform options.
  const TransformOptions({
    this.portal = ErrorPortal.localError,
    this.registry,
    this.defaultStatus = 500,
    this.defaultVersion = 'v1',
  });

  /// Portal used to build fallback type URIs.
  final ErrorPortal portal;

  /// Registry consulted when the value carries a known problem id; may be null.
  final ProblemRegistry? registry;

  /// Status for an uncatalogued fallback problem.
  final int defaultStatus;

  /// Version segment for fallback type URIs.
  final String defaultVersion;
}

/// Folds [error] into a typed [Problem].
///
/// - If [error] is already a [Problem], it is returned unchanged.
/// - If [options.registry] recognises an id carried on the value (a `problemId`
///   field on a Map, or a `problemId` getter), the registry's type/title/
///   recoverable are used and the `data` payload is preserved.
/// - Otherwise an uncatalogued problem (C0 §14) is produced: `type` built from
///   [TransformOptions.portal] with id [uncataloguedProblemId].
///
/// This never throws: the whole point is to guarantee a Problem for any value.
Problem fromObject(
  Object? error, {
  TransformOptions options = const TransformOptions(),
}) {
  if (error is Problem) {
    return error;
  }

  final ProblemRegistry? registry = options.registry;
  final String? id = _readProblemId(error);
  if (registry != null && id != null) {
    final ProblemType? type = registry.lookup(id);
    if (type != null) {
      return Problem(
        type: registry.typeUriFor(type),
        title: type.title,
        status: type.status ?? options.defaultStatus,
        recoverable: type.recoverable,
        data: _readData(error),
      );
    }
  }

  return Problem(
    type: problemTypeUri(
      portal: options.portal,
      version: options.defaultVersion,
      id: uncataloguedProblemId,
    ),
    title: 'Unexpected problem',
    status: options.defaultStatus,
    detail: error?.toString(),
    recoverable: false,
    data: const <String, Object?>{},
  );
}

String? _readProblemId(Object? error) {
  if (error is Map<String, Object?>) {
    final Object? id = error['problemId'];
    return id is String ? id : null;
  }
  // Ignore dynamic getters (avoid_dynamic_calls lint): only structured Maps
  // carry a declared problem id.
  return null;
}

Map<String, Object?> _readData(Object? error) {
  if (error is Map<String, Object?>) {
    final Object? data = error['data'];
    if (data is Map<Object?, Object?>) {
      return data.map(
        (Object? key, Object? value) => MapEntry(key.toString(), value),
      );
    }
  }
  return const <String, Object?>{};
}

File: tests/fixtures/api-baseline/lib/problem/transformer.go
package problem

import (
	"errors"
	"fmt"
)

// TransformOptions configures how [FromObject] folds an arbitrary value into a
// [Problem].
type TransformOptions struct {
	// Portal builds fallback type URIs for uncatalogued problems.
	Portal ErrorPortal
	// Registry, when non-nil, is consulted for values carrying a known
	// `problemId`.
	Registry *Registry
	// DefaultStatus is the status for an uncatalogued fallback problem.
	DefaultStatus int
	// DefaultVersion is the version segment for fallback type URIs.
	DefaultVersion string
}

// DefaultTransformOptions returns options that mint fallback URIs from
// [LocalErrorPortal] with a 500 status and the v1 version segment, and no
// registry.
func DefaultTransformOptions() TransformOptions {
	return TransformOptions{
		Portal:         LocalErrorPortal(),
		DefaultStatus:  defaultProblemStatus,
		DefaultVersion: defaultProblemVersion,
	}
}

// FromObject folds value into a typed [Problem]. It never panics: the whole
// point is to guarantee a Problem for any value.
//
//   - A [Problem] value is returned unchanged.
//   - An error carrying a [Error] anywhere in its chain (via errors.As)
//     yields that error's Problem.
//   - When options.Registry recognises a `problemId` carried on a
//     map[string]any value, the registry's type/title/status/recoverable are
//     used and the value's `data` map is preserved.
//   - Otherwise an uncatalogued problem (C0 §14) is produced with the
//     [UncataloguedProblemID] fallback type.
func FromObject(value any, options TransformOptions) Problem {
	if envelope, ok := value.(Problem); ok {
		return envelope
	}
	if err, ok := value.(error); ok {
		var problemErr *Error
		if errors.As(err, &problemErr) {
			return problemErr.Problem
		}
	}
	if options.Registry != nil {
		if id := problemIDFrom(value); id != "" {
			if problemType, ok := options.Registry.Lookup(id); ok {
				return registeredProblem(options, problemType, dataFrom(value))
			}
		}
	}
	return uncataloguedProblem(value, options)
}

func registeredProblem(options TransformOptions, problemType Type, data map[string]any) Problem {
	status := problemType.Status
	if status == 0 {
		status = options.DefaultStatus
	}
	typeURI, err := options.Registry.TypeURIFor(problemType)
	if err != nil {
		typeURI = blankType
	}
	return Problem{
		Type:        typeURI,
		Title:       problemType.Title,
		Status:      status,
		Recoverable: problemType.Recoverable,
		Data:        data,
	}
}

func uncataloguedProblem(value any, options TransformOptions) Problem {
	typeURI, err := TypeURI(options.Portal, options.DefaultVersion, UncataloguedProblemID)
	if err != nil {
		typeURI = blankType
	}
	return Problem{
		Type:        typeURI,
		Title:       unexpectedProblemTitle,
		Status:      options.DefaultStatus,
		Detail:      detailFrom(value),
		Recoverable: false,
		Data:        map[string]any{},
	}
}

func problemIDFrom(value any) string {
	if fields, ok := value.(map[string]any); ok {
		if id, ok := fields["problemId"].(string); ok {
			return id
		}
	}
	return ""
}

func dataFrom(value any) map[string]any {
	if fields, ok := value.(map[string]any); ok {
		if data, ok := fields["data"].(map[string]any); ok {
			return data
		}
	}
	return map[string]any{}
}

func detailFrom(value any) string {
	switch typed := value.(type) {
	case nil:
		return ""
	case error:
		return typed.Error()
	case string:
		return typed
	default:
		return fmt.Sprint(typed)
	}
}
